### PR TITLE
feat(change_detection): implement hydration/dehydration

### PR DESCRIPTION
**Pull request by vsavkin**

### CODE REVIEW

File: modules/angular2/src/change_detection/change_detection_jit_generator.es6 [hunk 1]
```diff
@@ -119,11 +119,22 @@ ${type}.prototype = Object.create(${ABSTRACT_CHANGE_DETECTOR}.prototype);
 `;
 }
 
-function setContextTemplate(type:string):string {
+function pipeOnDestroyTemplate(pipeNames:List) {
+  return pipeNames.map((p) => p + ".onDestroy()").join("\n");
```
Comment (rkirov) [inline modules/angular2/src/change_detection/change_detection_jit_generator.es6]: pet peeve: `'` vs `"` used inconsistently.

File: modules/angular2/src/change_detection/dynamic_change_detector.js [hunk 2]
```diff
@@ -43,15 +43,36 @@ export class DynamicChangeDetector extends AbstractChangeDetector {
...
     ListWrapper.fill(this.values, uninitialized);
     ListWrapper.fill(this.changes, false);
     ListWrapper.fill(this.pipes, null);
     ListWrapper.fill(this.prevContexts, uninitialized);
-    this.values[0] = context;
+  }
+
+  _destroyPipes() {
+    for(var i = 0; i < this.pipes.length; ++i) {
+      if (isPresent(this.pipes[i])) {
```
Comment (rkirov) [inline modules/angular2/src/change_detection/dynamic_change_detector.js]: OOC, why is the pipes array sparse?

File: modules/angular2/test/change_detection/change_detection_spec.js [hunk 3]
```diff
@@ -443,6 +443,35 @@ export function main() {
...
+            var cd = c["changeDetector"];
+
+            cd.hydrate("some context");
+            expect(cd.hydrated()).toBe(true);
+
+            cd.dehydrate();
+            expect(cd.hydrated()).toBe(false);
+
+            cd.hydrate("other context");
+            expect(cd.hydrated()).toBe(true);
```
Comment (rkirov) [inline modules/angular2/test/change_detection/change_detection_spec.js]: shouldn't you check that the values are cleared?

File: modules/angular2/src/change_detection/change_detection_jit_generator.es6 [hunk 4]
```diff
@@ -119,11 +119,22 @@ ${type}.prototype = Object.create(${ABSTRACT_CHANGE_DETECTOR}.prototype);
 `;
```
Comment (vicb) [inline modules/angular2/src/change_detection/change_detection_jit_generator.es6]: Would be great to update the generated code sample at the beginning of the file when this file is modified

File: modules/angular2/src/change_detection/change_detection_jit_generator.es6 (hunk 1 above)
Comment (vsavkin) [inline modules/angular2/src/change_detection/change_detection_jit_generator.es6]: Fixed

File: modules/angular2/src/change_detection/dynamic_change_detector.js (hunk 2 above)
Comment (vsavkin) [inline modules/angular2/src/change_detection/dynamic_change_detector.js]: so we can look them up by record self id

File: modules/angular2/src/change_detection/change_detection_jit_generator.es6 (hunk 4 above)
Comment (vsavkin) [inline modules/angular2/src/change_detection/change_detection_jit_generator.es6]: Fixed

**Pull request merged**